Writerside/topics/implementation/public-transit-service.md

```diff
@@ -1,9 +1,166 @@
+information about routes and upcoming departures.
+
+To optimize these operations, the `GtfsRaptorService` employs various indexing structures:
+
+1. **KDTree for Geospatial Queries**: To efficiently answer proximity queries (e.g., *"What are the nearest stops to my
+   location?"*), we use a k-dimensional tree (KDTree) that operates on geographic coordinates (latitude and longitude).
+   The distances between these coordinates are approximated using the Haversine formula [15], which provides sufficient
+   accuracy for most inhabited areas of the Earth (excluding extreme polar regions).
+
+2. **Compressed Trie for String Matching**: For searching stops by name, we implemented a compressed trie data
+   structure. This allows for efficient storage and retrieval of stop names based on prefix matching. The trie is filled
+   with all the stop names by inserting suffixes of the stop names into the structure [14]. This approach enables fast
+   lookups for queries like *"find stops whose names starts with, contains or ends with this substring."* Exmaple of how
+   the suffixTrie is filled:
+
+      ```java
+      public class SearchIndex<T> {
+   
+          // ...
+   
+          public void add(String key, T value) {
+   
+              for (int i = key.length() - 1; i >= 0; i--) {
+                  suffixTrie.insert(key.substring(i), entry);
+              }
+   
+          }
+   
+      }
+      ```
+
+Using these indices, queries for stops are highly efficient. The trade-off, however, is that these data structures
+consume more memory.
 
 ## Connection Routing Service
 
-RoutePartitioner
+The `ConnectionRoutingService` is responsible for finding connections between transit stops. In our `GtfsRaptorService`
+implementation, this service interacts with the [extended RAPTOR routing engine](extended-raptor.md), which processes
+routing requests based on the GTFS schedule data.
+
+### Route Partitioning
+
+The GTFS model and RAPTOR algorithm have differing perspectives on routes. In GTFS, a "route" refers to a user-friendly
+view of trips that may share a common line number or route name, even if the stop sequences differ slightly. For
+instance, a regional train might have trips that skip certain stops on later trips while still being part of the same
+route from a user's point of view.
+
+However, RAPTOR requires that all trips associated with a route must have the exact same stop sequence. This discrepancy
+is resolved by the `GtfsToRaptorConverter` by applying the `GtfsRoutePartitioner`, which splits GTFS routes into "
+sub-routes". These sub-routes group trips that share identical stop sequences, ensuring compatibility with RAPTOR’s
+routing model. During post-processing, when the routing results are returned to the user, the sub-routes are converted
+back into the original GTFS routes to maintain user-friendly responses (TODO: Where is this happening? in the RAPTOR or
```

Comment:
This is done in the Service this happens in the TypeMapper:

```
    public static Connection map(ch.naviqore.raptor.Connection connection, @Nullable Walk firstMile,
                                 @Nullable Walk lastMile, GtfsSchedule schedule) {
        List<Leg> legs = new ArrayList<>();

        if (firstMile != null) {
            legs.addFirst(firstMile);
        }

        for (ch.naviqore.raptor.Leg leg : connection.getLegs()) {
            legs.add(map(leg, schedule));
        }

        if (lastMile != null) {
            legs.addLast(lastMile);
        }

        return new GtfsRaptorConnection(legs);
    }

    public static Leg map(ch.naviqore.raptor.Leg leg, GtfsSchedule schedule) {
        int duration = (int) Duration.between(leg.getDepartureTime(), leg.getArrivalTime()).toSeconds();
        Stop sourceStop = map(schedule.getStops().get(leg.getFromStopId()));
        Stop targetStop = map(schedule.getStops().get(leg.getToStopId()));
        int distance = (int) Math.round(sourceStop.getLocation().distanceTo(targetStop.getLocation()));

        return switch (leg.getType()) {
            case WALK_TRANSFER ->
                    new GtfsRaptorTransfer(distance, duration, leg.getDepartureTime(), leg.getArrivalTime(), sourceStop,
                            targetStop);
            case ROUTE -> createPublicTransitLeg(leg, schedule, distance);
        };
    }
```

Reply:
https://github.com/naviqore/documentation/pull/30/commits/5f58b719276cad77e0558a29be950566a692c7c5